src/backend/distributed/planner/recursive_planning.c

```diff
@@ -321,12 +321,18 @@ RecursivelyPlanSubqueriesAndCTEs(Query *query, RecursivePlanningContext *context
 	/*
 	 * If the FROM clause is recurring (does not contain a distributed table),
 	 * then we cannot have any distributed tables appearing in subqueries in
-	 * the WHERE clause.
+	 * the SELECT and WHERE clauses.
 	 */
-	if (ShouldRecursivelyPlanAllSubqueriesInWhere(query))
+	if (ShouldRecursivelyPlanSublinks(query))
 	{
 		/* replace all subqueries in the WHERE clause */
-		RecursivelyPlanAllSubqueries((Node *) query->jointree->quals, context);
+		if (query->jointree && query->jointree->quals)
+		{
+			RecursivelyPlanAllSubqueries((Node *) query->jointree->quals, context);
+		}
+
+		/* replace all subqueries in the SELECT clause */
+		RecursivelyPlanAllSubqueries((Node *) query->targetList, context);
```

Comment:
Why do we recursively plan everything there, wouldn't that be enough to only recursively plan subqueries that have a distributed table?

![image](https://user-images.githubusercontent.com/17179875/104457868-a3c87200-55bb-11eb-8256-a5f4d3febb00.png)
(`a` is a local table)
For example, here do we need to recursively plan `(SELECT * FROM a)`?

Reply:
I noticed that too, and looked into it briefly, but it seemed beyond the scope of this PR.

We should add more checks to RecursivelyPlanAllSubqueries to address it for both WHERE and SELECT.

Reply:
Yeah I agree it is beyond the scope of this PR. I was mostly curious if there was any specific reason, but seems like no :)

Reply:
@marcocitus I think this is not because of `RecursivelyPlanAllSubqueries`, but because of `ShouldRecursivelyPlanSubquery`. Why don't we return false here? https://github.com/citusdata/citus/blob/b79af16dac6a738703cfb315ab4595a690ec50a1/src/backend/distributed/planner/recursive_planning.c#L908-L920